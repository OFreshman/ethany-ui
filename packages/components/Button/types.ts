import type { Component } from "vue";

export type ButtonType = "success" | "primary" | "warning" | "danger" | "info";
export type NativeType = "button" | "reset" | "submit";
export type ButtonSize = "large" | "default" | "small";
export interface ButtonProps {
  tag?: Component;
  type?: ButtonType;
  size?: ButtonSize;
  nativeType?: NativeType;
  disabled?: boolean;
  loading?: boolean;
  icon?: string;
  circle?: boolean;
  plain?: boolean;
  round?: boolean;
}
