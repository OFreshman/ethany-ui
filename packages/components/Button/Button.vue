<script setup lang="ts">
/**
 * 组件目录结构
 * Button.vue  组件
 * Button.test.tsc  测试文件
 * type.ts 类型声明
 * style.css 样式文件
 * constants.ts 常量（如果有)
 */
defineOptions({
  name: "EyButton",
  // props: {
  //   // disabled: Boolean,
  // },
});
</script>
<template>
  <button style="background-color: blue; color: red">i am a button</button>
</template>
