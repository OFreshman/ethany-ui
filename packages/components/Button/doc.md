# 需求分析与功能点设计

## 需求分析 [XQFX]

### 用户调研摘要

通过对现有用户的调研，我们发现用户在使用按钮组件时有以下几个主要需求：

1. **多样化的按钮样式**：用户希望能够使用不同类型、形状和尺寸的按钮，以适应不同的使用场景。
2. **图标按钮**：用户希望按钮能够包含图标，以便在视觉上更直观地传达功能。
3. **禁用状态**：用户希望能够禁用按钮，以防止误操作。
4. **加载状态**：用户希望按钮能够显示加载状态，以便在执行异步操作时提供反馈。
5. **按钮组**：用户希望能够将多个按钮组合在一起，以便在界面上更紧凑地展示相关操作。

### 竞品对比报告

在对比了市场上几款流行的 UI 组件库（如 Element UI、Ant Design、Vuetify）的按钮组件后，我们发现：

1. **Element UI**：提供了丰富的按钮样式和状态支持，包括不同类型、形状、尺寸、图标按钮、禁用状态、加载状态和按钮组。
2. **Ant Design**：同样提供了丰富的按钮样式和状态支持，并且在交互设计上更加注重细节，例如按钮的点击反馈和加载动画。
3. **Vuetify**：提供了多样化的按钮样式和状态支持，并且在响应式设计上表现出色，能够适应不同屏幕尺寸的设备。

### 市场趋势分析

随着前端技术的发展，用户对 UI 组件的要求越来越高，尤其是对按钮组件的需求更加多样化和个性化。未来的趋势可能包括：

1. **更丰富的样式和状态支持**：用户希望能够自定义按钮的样式和状态，以便更好地适应不同的使用场景。
2. **更好的交互体验**：用户希望按钮在交互上能够提供更好的反馈，例如点击动画、加载动画等。
3. **响应式设计**：用户希望按钮能够在不同屏幕尺寸的设备上表现出色，提供一致的用户体验。

### 用户痛点

1. **样式单一**：现有的按钮样式较为单一，无法满足用户对多样化样式的需求。
2. **缺乏图标支持**：现有的按钮组件缺乏图标支持，无法在视觉上直观地传达功能。
3. **缺乏加载状态**：现有的按钮组件缺乏加载状态，无法在执行异步操作时提供反馈。
4. **缺乏按钮组**：现有的按钮组件无法将多个按钮组合在一起，导致界面上操作按钮分散，不够紧凑。

### 期望功能

1. **多样化的按钮样式**：支持不同类型（如 primary、success、warning、danger、info）、形状（如 plain、round、circle）和尺寸（如 large、default、small）的按钮。
2. **图标按钮**：支持在按钮中添加图标，以便在视觉上更直观地传达功能。
3. **禁用状态**：支持禁用按钮，以防止误操作。
4. **加载状态**：支持按钮显示加载状态，以便在执行异步操作时提供反馈。
5. **按钮组**：支持将多个按钮组合在一起，以便在界面上更紧凑地展示相关操作。

### 安全性需求

1. **防止误操作**：通过禁用状态和节流模式，防止用户在按钮执行操作时误操作。
2. **数据安全**：确保按钮在执行操作时不会泄露用户数据。

## 功能点设计 [GNSJ]

### 功能描述

Eric-UI 的按钮组件提供了多种类型、形状和尺寸的按钮，支持图标、禁用状态、加载状态和按钮组功能，以满足用户在不同场景下的需求。

### API 设计

#### Button 组件

##### Props

| Name              | Description                         | Type                   | Default   |
| ----------------- | ----------------------------------- | ---------------------- | --------- | --------- | --------- | ------- | --------- |
| size              | 按钮尺寸                            | `enum` - `'large'      | 'default' | 'small'`  | 'default' |
| type              | 按钮类型                            | `enum` - `'primary'    | 'success' | 'warning' | 'danger'  | 'info'` | 'default' |
| plain             | 是否为朴素按钮                      | `boolean`              | false     |
| round             | 是否为圆角按钮                      | `boolean`              | false     |
| circle            | 是否为圆形按钮                      | `boolean`              | false     |
| loading           | 是否为加载中状态                    | `boolean`              | false     |
| loading-icon      | 自定义加载中状态图标组件            | `string`               | 'spinner' |
| disabled          | 按钮是否为禁用状态                  | `boolean`              | false     |
| icon              | 按钮图标                            | `string`               | -         |
| autofocus         | 是否自动聚焦(原生 `autofocus` 属性) | `boolean`              | false     |
| native-type       | 原生 type 属性                      | `enum` - `'button'     | 'submit'  | 'reset'`  | 'button'  |
| tag               | 自定义元素标签                      | `string` / `Component` | 'button'  |
| use-throttle      | 是否使用节流模式                    | `boolean`              | true      |
| throttle-duration | 节流模式下，节流时间间隔(ms)        | `number`               | 500       |

##### Events

| Name  | Description  | Type                          |
| ----- | ------------ | ----------------------------- |
| click | 按钮点击事件 | `(event: MouseEvent) => void` |

##### Slots

| Name    | Description        |
| ------- | ------------------ |
| default | 默认插槽, 按钮内容 |
| loading | 自定义加载图标     |

##### Expose

| Name     | Description    | Type                     |
| -------- | -------------- | ------------------------ | --------- | --------- |
| ref      | 按钮 html 元素 | `Ref<HTMLButtonElement>` |
| size     | 按钮尺寸       | `ComputedRef<''          | 'small'   | 'large'>` |
| type     | 按钮类型       | `ComputedRef<''          | 'primary' | ...>`     |
| disabled | 按钮禁用状态   | `ComputedRef<boolean>`   |

#### ButtonGroup 组件

##### Props

| Name     | Description          | Type                | Default   |
| -------- | -------------------- | ------------------- | --------- | --------- | --------- | ------- | --------- |
| size     | 按钮组尺寸           | `enum` - `'large'   | 'default' | 'small'`  | 'default' |
| type     | 按钮组类型           | `enum` - `'primary' | 'success' | 'warning' | 'danger'  | 'info'` | 'default' |
| disabled | 按钮组是否为禁用状态 | `boolean`           | false     |

##### Slots

| Name    | Description | Sub Component |
| ------- | ----------- | ------------- |
| default | 默认插槽    | Button        |

### 交互关系

1. **按钮点击**：用户点击按钮时，会触发 `click` 事件，并根据按钮的 `type`、`disabled` 等属性决定按钮的行为。
2. **加载状态**：当按钮处于加载状态时，按钮会显示加载图标，并且禁用点击操作。
3. **禁用状态**：当按钮处于禁用状态时，按钮会显示为灰色，并且禁用点击操作。
4. **按钮组**：按钮组内的按钮会共享相同的 `size` 和 `type` 属性，并且可以一起禁用。

### 功能实现细节

1. **多样化的按钮样式**：通过 `type`、`plain`、`round` 和 `circle` 属性来定义按钮的样式。
2. **图标按钮**：通过 `icon` 属性来定义按钮的图标。
3. **禁用状态**：通过 `disabled` 属性来定义按钮是否被禁用。
4. **加载状态**：通过 `loading` 和 `loading-icon` 属性来定义按钮的加载状态。
5. **按钮组**：通过 `ButtonGroup` 组件来对多个按钮进行分组。

### 用户操作流程

1. **点击按钮**：用户点击按钮，触发 `click` 事件。
2. **加载状态**：如果按钮处于加载状态，显示加载图标，并禁用点击操作。
3. **禁用状态**：如果按钮处于禁用状态，显示为灰色，并禁用点击操作。
4. **按钮组操作**：用户点击按钮组内的按钮，触发相应的 `click` 事件。

### 异常处理

1. **按钮点击无响应**：检查按钮是否处于禁用状态或加载状态。
2. **加载图标显示异常**：检查 `loading-icon` 属性是否设置正确。
3. **按钮组操作异常**：检查 `ButtonGroup` 组件的 `size` 和 `type` 属性是否设置正确。
