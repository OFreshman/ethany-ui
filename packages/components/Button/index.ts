import Button from "./Button.vue";
import { withInstall } from "@ethany-ui/utils";

export const EyButton = withInstall(Button);
