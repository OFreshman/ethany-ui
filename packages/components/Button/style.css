.ey-button-group {
  --ey-button-group-border-color: var(--ey-border-color-lighter);
}
.ey-button {
  --ey-button-font-weight: var(--ey-font-weight-primary);
  --ey-button-border-color: var(--ey-border-color);
  --ey-button-bg-color: var(--ey-fill-color-blank);
  --ey-button-text-color: var(--ey-text-color-regular);
  --ey-button-disabled-text-color: var(--ey-disabled-text-color);
  --ey-button-disabled-bg-color: var(--ey-fill-color-blank);
  --ey-button-disabled-border-color: var(--ey-border-color-light);
  --ey-button-hover-text-color: var(--ey-color-primary);
  --ey-button-hover-bg-color: var(--ey-color-primary-light-9);
  --ey-button-hover-border-color: var(--ey-color-primary-light-7);
  --ey-button-active-text-color: var(--ey-button-hover-text-color);
  --ey-button-active-border-color: var(--ey-color-primary);
  --ey-button-active-bg-color: var(--ey-button-hover-bg-color);
  --ey-button-outline-color: var(--ey-color-primary-light-5);
  --ey-button-active-color: var(--ey-text-color-primary);
}

.ey-button {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  line-height: 1;
  height: 32px;
  white-space: nowrap;
  cursor: pointer;
  color: var(--ey-button-text-color);
  text-align: center;
  box-sizing: border-box;
  outline: none;
  transition: 0.1s;
  font-weight: var(--ey-button-font-weight);
  user-select: none;
  vertical-align: middle;
  -webkit-appearance: none;
  background-color: var(--ey-button-bg-color);
  border: var(--ey-border);
  border-color: var(--ey-button-border-color);
  padding: 8px 15px;
  font-size: var(--ey-font-size-base);
  border-radius: var(--ey-border-radius-base);
  & + & {
    margin-left: 12px;
  }
  &:hover,
  &:focus {
    color: var(--ey-button-hover-text-color);
    border-color: var(--ey-button-hover-border-color);
    background-color: var(--ey-button-hover-bg-color);
    outline: none;
  }
  &:active {
    color: var(--ey-button-active-text-color);
    border-color: var(--ey-button-active-border-color);
    background-color: var(--ey-button-active-bg-color);
    outline: none;
  }
  /*plain*/
  &.is-plain {
    --ey-button-hover-text-color: var(--ey-color-primary);
    --ey-button-hover-bg-color: var(--ey-fill-color-blank);
    --ey-button-hover-border-color: var(--ey-color-primary);
  }
  /*round*/
  &.is-round {
    border-radius: var(--ey-border-radius-round);
  }
  /*circle*/
  &.is-circle {
    border-radius: 50%;
    padding: 8px;
  }
  /*disabled*/
  &.is-loading,
  &.is-disabled,
  &.is-disabled:hover,
  &.is-disabled:focus,
  &[disabled],
  &[disabled]:hover,
  &[disabled]:focus {
    color: var(--ey-button-disabled-text-color);
    cursor: not-allowed;
    background-image: none;
    background-color: var(--ey-button-disabled-bg-color);
    border-color: var(--ey-button-disabled-border-color);
  }
  [class*="ey-icon"] {
    width: 1em;
    height: 1em;
  }
}
@each $val in primary, success, warning, info, danger {
  .ey-button--$(val) {
    --ey-button-text-color: var(--ey-color-white);
    --ey-button-bg-color: var(--ey-color-$(val));
    --ey-button-border-color: var(--ey-color-$(val));
    --ey-button-outline-color: var(--ey-color-$(val)-light-5);
    --ey-button-active-color: var(--ey-color-$(val)-dark-2);
    --ey-button-hover-text-color: var(--ey-color-white);
    --ey-button-hover-bg-color: var(--ey-color-$(val)-light-3);
    --ey-button-hover-border-color: var(--ey-color-$(val)-light-3);
    --ey-button-active-bg-color: var(--ey-color-$(val)-dark-2);
    --ey-button-active-border-color: var(--ey-color-$(val)-dark-2);
    --ey-button-disabled-text-color: var(--ey-color-white);
    --ey-button-disabled-bg-color: var(--ey-color-$(val)-light-5);
    --ey-button-disabled-border-color: var(--ey-color-$(val)-light-5);
  }
  .ey-button--$(val).is-plain {
    --ey-button-text-color: var(--ey-color-$(val));
    --ey-button-bg-color: var(--ey-color-$(val)-light-9);
    --ey-button-border-color: var(--ey-color-$(val)-light-5);
    --ey-button-hover-text-color: var(--ey-color-white);
    --ey-button-hover-bg-color: var(--ey-color-$(val));
    --ey-button-hover-border-color: var(--ey-color-$(val));
    --ey-button-active-text-color: var(--ey-color-white);

    --ey-button-disabled-text-color: var(--ey-color-$(val)-light-5);
    --ey-button-disabled-bg-color: var(--ey-color-$(val)-light-9);
    --ey-button-disabled-border-color: var(--ey-color-$(val)-light-8);
  }
}
.ey-button--large {
  --ey-button-size: 40px;
  height: var(--ey-button-size);
  padding: 12px 19px;
  font-size: var(--ey-font-size-base);
  bordey-radius: var(--ey-bordey-radius-base);
  /*circle*/
  &.is-circle {
    border-radius: 50%;
    padding: 12px;
  }
}
.ey-button--small {
  --ey-button-size: 24px;
  height: var(--ey-button-size);
  padding: 5px 11px;
  font-size: 12px;
  border-radius: calc(var(--ey-border-radius-base) - 1px);
  /*circle*/
  &.is-circle {
    border-radius: 50%;
    padding: 5px;
  }
  [class*="ey-icon"] {
    width: 12px;
    height: 12px;
  }
}

.ey-button-group {
  display: inline-block;
  vertical-align: middle;

  &::after {
    clear: both;
  }
  & > :deep(ey-button) {
    float: left;
    position: relative;
    margin-left: 0;
    &:first-child {
      border-top-right-radius: 0;
      border-bottom-right-radius: 0;
      border-right-color: var(--ey-button-group-border-color);
    }
    &:last-child {
      border-top-left-radius: 0;
      border-bottom-left-radius: 0;
      border-left-color: var(--ey-button-group-border-color);
    }
    &:not(:first-child):not(:last-child) {
      border-radius: 0;

      border-left-color: var(--ey-button-group-border-color);
      border-right-color: var(--ey-button-group-border-color);
    }
    &:not(:last-child) {
      margin-right: -1px;
    }
    &:first-child:last-child {
      border-top-right-radius: var(--ey-border-radius-base);
      border-bottom-right-radius: var(--ey-border-radius-base);
      border-top-left-radius: var(--ey-border-radius-base);
      border-bottom-left-radius: var(--ey-border-radius-base);

      &.is-round {
        border-radius: var(--ey-border-radius-round);
      }

      &.is-circle {
        border-radius: 50%;
      }
    }
  }
}
