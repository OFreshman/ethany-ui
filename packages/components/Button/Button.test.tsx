import { describe, it, expect, vi } from "vitest";
import { mount } from "@vue/test-utils";
// import { defineComponent } from 'vue';
import Button from "./Button.vue"; // 根据实际路径调整

// 定义一个用于测试的组件
// const TestButton = defineComponent({
//   components: {
//     Button,
//   },
//   template: `
//     <Button
//       :size="size"
//       :type="type"
//       :plain="plain"
//       :round="round"
//       :circle="circle"
//       :loading="loading"
//       :disabled="disabled"
//       :icon="icon"
//       :autofocus="autofocus"
//       :native-type="nativeType"
//       :use-throttle="useThrottle"
//       :throttle-duration="throttleDuration"
//       @click="onClick"
//     >
//       <template v-if="$slots.default" #default>{{ content }}</template>
//       <template v-if="$slots.loading" #loading>{{ loadingContent }}</template>
//     </Button>
//   `,
//   props: {
//     size: String,
//     type: String,
//     plain: Boolean,
//     round: Boolean,
//     circle: Boolean,
//     loading: Boolean,
//     disabled: Boolean,
//     icon: String,
//     autofocus: Boolean,
//     nativeType: String,
//     useThrottle: Boolean,
//     throttleDuration: Number,
//     content: String,
//     loadingContent: String,
//   },
//   setup(props, { emit }) {
//     const onClick = vi.fn();
//     return { ...props, onClick };
//   },
// });

describe("Button Component", () => {
  it("should render correctly with default props", () => {
    const wrapper = mount(Button, {
      props: {
        content: "Click Me",
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render correctly with different sizes", () => {
    const sizes = ["large", "default", "small"];
    sizes.forEach((size) => {
      const wrapper = mount(Button, {
        props: {
          size,
          content: `Button with size ${size}`,
        },
      });
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  it("should render correctly with different types", () => {
    const types = ["primary", "success", "warning", "danger", "info"];
    types.forEach((type) => {
      const wrapper = mount(Button, {
        props: {
          type,
          content: `Button with type ${type}`,
        },
      });
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  it("should render icon when icon prop is provided", () => {
    const wrapper = mount(Button, {
      props: {
        icon: "check",
        content: "Icon Button",
      },
    });
    expect(wrapper.find(".icon").exists()).toBe(true);
  });

  // it("should be disabled when disabled prop is true", () => {
  //   const wrapper = mount(Button, {
  //     props: {
  //       disabled: true,
  //       content: "Disabled Button",
  //     },
  //   });
  //   expect(wrapper.element).toBeDisabled();
  // });

  it("should be disabled when the disabled prop is true", () => {
    const wrapper = mount(Button, {
      props: {
        disabled: true,
      },
    });
    expect(wrapper.element.disabled).toBe(true);
  });

  it("should be in loading state when loading prop is true", () => {
    const wrapper = mount(Button, {
      props: {
        loading: true,
        content: "Loading Button",
      },
    });
    expect(wrapper.find(".loading").exists()).toBe(true);
  });

  // it('should emit click event on button click', async () => {
  //   const wrapper = mount(Button, {
  //     props: {
  //       content: 'Click Me',
  //     },
  //   });
  //   await wrapper.trigger('click');
  //   expect(wrapper.vm.onClick).toHaveBeenCalled();
  // });

  // it('should not emit click event when disabled', async () => {
  //   const wrapper = mount(Button, {
  //     props: {
  //       disabled: true,
  //       content: 'Do Not Click',
  //     },
  //   });
  //   await wrapper.trigger('click');
  //   expect(wrapper.vm.onClick).not.toHaveBeenCalled();
  // });

  // Add more test cases as needed to cover other scenarios and functionalities.
});
