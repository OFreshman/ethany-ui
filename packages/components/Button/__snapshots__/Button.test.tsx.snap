// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > should render correctly with default props 1`] = `"<button style="background-color: blue; color: red;" content="Click Me">i am a button</button>"`;

exports[`Button Component > should render correctly with different sizes 1`] = `"<button style="background-color: blue; color: red;" size="large" content="Button with size large">i am a button</button>"`;

exports[`Button Component > should render correctly with different sizes 2`] = `"<button style="background-color: blue; color: red;" size="default" content="Button with size default">i am a button</button>"`;

exports[`Button Component > should render correctly with different sizes 3`] = `"<button style="background-color: blue; color: red;" size="small" content="Button with size small">i am a button</button>"`;

exports[`Button Component > should render correctly with different types 1`] = `"<button style="background-color: blue; color: red;" type="submit" content="Button with type primary">i am a button</button>"`;

exports[`Button Component > should render correctly with different types 2`] = `"<button style="background-color: blue; color: red;" type="submit" content="Button with type success">i am a button</button>"`;

exports[`Button Component > should render correctly with different types 3`] = `"<button style="background-color: blue; color: red;" type="submit" content="Button with type warning">i am a button</button>"`;

exports[`Button Component > should render correctly with different types 4`] = `"<button style="background-color: blue; color: red;" type="submit" content="Button with type danger">i am a button</button>"`;

exports[`Button Component > should render correctly with different types 5`] = `"<button style="background-color: blue; color: red;" type="submit" content="Button with type info">i am a button</button>"`;

exports[`Button Componentvue > should render in disabled state 1`] = `
"<button style="background-color: blue; color: red;" disabled="">i am a button</button>
<button style="background-color: blue; color: red;" type="submit" disabled="">i am a button</button>
<button style="background-color: blue; color: red;" type="submit" disabled="">i am a button</button>"
`;

exports[`Button Componentvue > should render in loading state 1`] = `
"<button style="background-color: blue; color: red;" loading="true">i am a button</button>
<button style="background-color: blue; color: red;" type="submit" loading="true">i am a button</button>
<button style="background-color: blue; color: red;" type="submit" loading="true">i am a button</button>"
`;

exports[`Button Componentvue > should render with different shapes 1`] = `
"<button style="background-color: blue; color: red;" plain="true">i am a button</button>
<button style="background-color: blue; color: red;" round="true">i am a button</button>
<button style="background-color: blue; color: red;" circle="true">i am a button</button>"
`;

exports[`Button Componentvue > should render with different sizes 1`] = `
"<button style="background-color: blue; color: red;" size="large">i am a button</button>
<button style="background-color: blue; color: red;">i am a button</button>
<button style="background-color: blue; color: red;" size="small">i am a button</button>"
`;

exports[`Button Componentvue > should render with different types 1`] = `
"<button style="background-color: blue; color: red;" type="submit">i am a button</button>
<button style="background-color: blue; color: red;" type="submit">i am a button</button>
<button style="background-color: blue; color: red;" type="submit">i am a button</button>
<button style="background-color: blue; color: red;" type="submit">i am a button</button>
<button style="background-color: blue; color: red;" type="submit">i am a button</button>"
`;

exports[`Button Componentvue > should render with icons 1`] = `
"<button style="background-color: blue; color: red;" icon="search">i am a button</button>
<button style="background-color: blue; color: red;" icon="edit">i am a button</button>
<button style="background-color: blue; color: red;" icon="check">i am a button</button>"
`;
